### PR TITLE
Per bug 1377566, OpenRaster format improvements, and OpenDocument standard.

### PR DESCRIPTION
**Pull request by horkana**
includes fix for opacity/visibilty;
enhancements: selected layer; mergedimage; OpenDocument compatibility;

### CODE REVIEW

File: Pinta.Core/Actions/HelpActions.cs [hunk 1]
```diff
@@ -47,10 +47,10 @@ public HelpActions ()
 			fact.Add ("Menu.Help.Translate.png", new Gtk.IconSet (PintaCore.Resources.GetIcon ("Menu.Help.Translate.png")));
 			fact.AddDefault ();
 
-			Website = new Gtk.Action ("Website", Catalog.GetString ("Pinta Website"), null, "Menu.Help.Website.png");
-			Bugs = new Gtk.Action ("Bugs", Catalog.GetString ("File a Bug"), null, "Menu.Help.Bug.png");
-			Translate = new Gtk.Action ("Translate", Catalog.GetString ("Translate This Application"), null, "Menu.Help.Translate.png");
-			About = new Gtk.Action ("About", Catalog.GetString ("About"), null, Stock.About);
+			Website = new Gtk.Action ("Website", string.Format("{0} {1}", PintaCore.ApplicationName, Catalog.GetString ("_Website")), null, "Menu.Help.Website.png");
```
Comment (cameronwhite) [inline Pinta.Core/Actions/HelpActions.cs]: The format string needs to be translatable (`string.Format (Catalog.GetString ("{0} _Website"))`) since grammar rules can be very different in other languages.

You may want to place a comment (prefixed with "Translators:" so that our automated tools pick it up) on the preceding line to explain to translators that the application name will be substituted there, since the meaning of the string may be confusing otherwise.

File: Pinta.Core/Classes/Document.cs [hunk 2]
```diff
@@ -40,13 +40,20 @@ namespace Pinta.Core
 	// somewhat arbitrary.  In general:
 	// Document - Data about the image itself
 	// Workspace - Data about Pinta's state for the image
-	public class Document
+	public class Document : ObservableObject
```
Comment (cameronwhite) [inline Pinta.Core/Classes/Document.cs]: I'm not clear why `Document` needs to derive from `ObservableObject` here?

File: Pinta.Core/Classes/Document.cs [hunk 3]
```diff
@@ -40,13 +40,20 @@ namespace Pinta.Core
...
 	{
 		private string filename;
 		private bool is_dirty;
 		private int layer_name_int = 2;
 		private int current_layer = -1;
 
+        // metadata
+        private string author;
```
Comment (cameronwhite) [inline Pinta.Core/Classes/Document.cs]: Why not just store a `DocumentProperties` member here?

File: Pinta.Core/Classes/Document.cs (hunk 2 above)
Comment (horkana) [inline Pinta.Core/Classes/Document.cs]: It is difficult to remember after this long. 
It might not be necessary anymore. 

Initially the dialog had non-modal instant apply behaviour, which wasn't what I wanted.

File: Pinta.Core/Classes/Document.cs (hunk 3 above)
Comment (horkana) [inline Pinta.Core/Classes/Document.cs]: A lot of this code was based on the example of other code seen in Pinta. Also since I later made the dialog into a simple blocking/modal dialog it may not be necessary at all.

File: Pinta.Core/Classes/Document.cs (hunk 2 above)
Comment (cameronwhite) [inline Pinta.Core/Classes/Document.cs]: OK, I'd guess it can probably be removed - if nothing anywhere is subscribing to the PropertyChanged even, then deriving from ObservableObject isn't necessary.

File: Pinta.Core/Actions/HelpActions.cs [hunk 4]
```diff
@@ -47,10 +47,11 @@ public HelpActions ()
 			fact.Add ("Menu.Help.Translate.png", new Gtk.IconSet (PintaCore.Resources.GetIcon ("Menu.Help.Translate.png")));
 			fact.AddDefault ();
 
-			Website = new Gtk.Action ("Website", Catalog.GetString ("Pinta Website"), null, "Menu.Help.Website.png");
-			Bugs = new Gtk.Action ("Bugs", Catalog.GetString ("File a Bug"), null, "Menu.Help.Bug.png");
-			Translate = new Gtk.Action ("Translate", Catalog.GetString ("Translate This Application"), null, "Menu.Help.Translate.png");
-			About = new Gtk.Action ("About", Catalog.GetString ("About"), null, Stock.About);
+			// Translators: "Application name" (Pinta) + "Website"
```
Comment (cameronwhite) [inline Pinta.Core/Actions/HelpActions.cs]: Thanks for adding the comment for translators.
The format string itself still needs to be made translatable, though - see my previous comment.for what it should look like

File: Pinta.Core/Classes/Layer.cs [hunk 5]
```diff
@@ -77,15 +80,20 @@ public Layer (ImageSurface surface, bool hidden, double opacity, string name)
 			set { if (hidden != value) SetValue (HiddenProperty, ref hidden, value); }
 		}
 		
+		public bool Locked {
```
Comment (cameronwhite) [inline Pinta.Core/Classes/Layer.cs]: Does locking a layer actually do anything currently? If not, I'd suggest filing a bug so that we don't lose track of it before the 1.7 release (adding layer locking functionality would be preferable as a separate pull request, anyways)

File: Pinta/Dialogs/DocumentPropertiesDialog.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            table.Attach(keywordsLabel, 0, 1, 3, 4, AttachOptions.Fill, 0, xPad, yPad);
+            
+            keywordsEntry = new Entry ();
+            keywordsLabel.MnemonicWidget = keywordsEntry;
+            table.Attach(keywordsEntry, 1, 2, 3, 4, AttachOptions.Fill, 0, xPad, yPad);
+
+            Label commentsLabel = new Label (Mono.Unix.Catalog.GetString ("Co_mments:"));
+            table.Attach(commentsLabel, 0, 1, 4, 5, AttachOptions.Fill, 0, xPad, yPad);
+            
+            // TODO ideally this would be a scrolled box
```
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: I think Pinta's error dialog would provide an example of a scrollable text field.

File: Pinta/Dialogs/DocumentPropertiesDialog.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+using Mono.Unix;
+using Pinta.Core;
+
+namespace Pinta
+{
+    public class DocumentPropertiesDialog : Dialog
+    {
+        private DocumentProperties initial_properties;
+
+        public string author;
```
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Nitpick: the usual convention is for public fields to be named in `UpperCase`

File: Pinta/Dialogs/DocumentPropertiesDialog.cs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        public string title;
+        public string subject;
+        public string keywords;
+        public string comments;
+
+        private Entry authorEntry;
```
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Nitpick: Pinta's source code is fairly inconsistent about how private members are named, but it should at least be consistent within a class - `lower_case` is what I prefer

File: Pinta/Dialogs/DocumentPropertiesDialog.cs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        private Entry keywordsEntry;
+        private TextView commentsTextView; 
+        
+        public DocumentPropertiesDialog () : base (Catalog.GetString ("Properties"), PintaCore.Chrome.MainWindow, DialogFlags.Modal, Stock.Cancel, ResponseType.Cancel, Stock.Ok, ResponseType.Ok)
+        {
+            Build ();
+
+            this.Icon = PintaCore.Resources.GetIcon (Stock.Properties, 16);
+            
+            // suggest current user as author name
```
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Are these intentionally left as commented out?

File: Pinta/Dialogs/DocumentPropertiesDialog.cs (hunk 7 above)
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Actually, why not just store an `updated_properties` member?

File: Pinta/Dialogs/DocumentPropertiesDialog.cs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+            comments = commentsTextView.Buffer.Text;
+        }
+        
+        private void Build ()
+        {
+            DefaultWidth = 400;
+            DefaultHeight = 300;
+            BorderWidth = 6;
+
+            VBox tBox = new VBox ();// HACK handwriting GUI code sucks
```
Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: I'm not clear why is this needed, versus just using the VBox property of the dialog?

File: Pinta/Dialogs/AboutDialog.cs [hunk 11]
```diff
@@ -274,11 +274,11 @@ internal class AboutDialog : Dialog
 
 		public AboutDialog () : base (string.Empty, PintaCore.Chrome.MainWindow, DialogFlags.Modal)
 		{
-			Title = Catalog.GetString ("About Pinta");
+			Title = string.Format ("{0} {1}", Catalog.GetString ("About"), PintaCore.ApplicationName);
```
Comment (cameronwhite) [inline Pinta/Dialogs/AboutDialog.cs]: As with my other comment, the format string needs to be translatable.

File: Pinta/Actions/File/PropertiesAction.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                );
+
+                var historyItem = new SimpleHistoryItem (
+                    Stock.Properties,
+                    historyMessage
+                );
+                
+                PintaCore.Workspace.ActiveWorkspace.History.PushNewItem (historyItem);
+
+                PintaCore.Workspace.ActiveWorkspace.Invalidate ();
```
Comment (cameronwhite) [inline Pinta/Actions/File/PropertiesAction.cs]: This forces a redraw of the canvas - I don't think that's necessary here?

File: Pinta/Actions/File/PropertiesAction.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+            } else {
+				// Cancel was pressed, reset dialog to initial properties
+                var initial = dialog.InitialDocumentProperties;
```
Comment (cameronwhite) [inline Pinta/Actions/File/PropertiesAction.cs]: Is it necessary to do anything with the dialog here? It's about to be destroyed

Comment (horkana) [inline Pinta/Actions/File/PropertiesAction.cs]: Again I think it was from before it was modal, that could probably go too.

File: Pinta/Actions/File/PropertiesAction.cs (hunk 12 above)
Comment (horkana) [inline Pinta/Actions/File/PropertiesAction.cs]: I think it was from before it was modal, that could probably go.

File: Pinta/Dialogs/DocumentPropertiesDialog.cs (hunk 10 above)
Comment (horkana) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Gtk box model is horrible and confusing, so it got a bit messy, I regret not using glade or stetic etc. 
It is inevitable that this dialogue will expand, so I feel the tabs are a necessity. I had intended tBox to be a separate method to draw the contents of the tab page, but the code restructuring wasn't so much a priority as getting it working and looking better for the user.

File: Pinta/Dialogs/DocumentPropertiesDialog.cs (hunk 9 above)
Comment (horkana) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: Basically yes. 
It more than a little creepy having these filled in automatically. 
I had a go at suggestion completion but it was an ugly mess and deleted it, but it did seem like a less invasive way to give the user some helpful info.

File: Pinta.Core/Classes/Layer.cs (hunk 5 above)
Comment (horkana) [inline Pinta.Core/Classes/Layer.cs]: I wrote this essentially so that it would pass through any locking information if it was set. Probably best to comment out the checkbox in the dialog. 

Locking can be confusing for users if it is is not made very very clear why they can't do what they are trying to do (even having a menu greyed out can be confusing, and not obvious how to get to the feature). 

Lots of these little things got written while I was taking a break from trying to figure out the things I couldn't get to work.

File: Pinta.Core/Actions/HelpActions.cs (hunk 4 above)
Comment (horkana) [inline Pinta.Core/Actions/HelpActions.cs]: I've misunderstood what you were asking for here. 

Tried again but and thought I almost had it (something to do with Catalog.GetString marking it as a translatable string but I'm not getting it quite right) but will need to leave it until I've more time.

**Comment by horkana** [comment]
I don't expect I'll have time to look at this again before the weekend (definitely not time to rewrite anything, maybe time to read and comment here, and read the code maybe). 

Thanks for your comments.

File: Pinta.Core/Actions/HelpActions.cs (hunk 4 above)
Comment (cameronwhite) [inline Pinta.Core/Actions/HelpActions.cs]: No problem, this isn't well-documented - I've added a wiki page (https://github.com/PintaProject/Pinta/wiki/Translations) about how translations work, which hopefully helps

**Comment by cameronwhite** [comment]
OK, no problem. I think I've reviewed everything except for the .ora format changes, so I'll try to look at those by the weekend.

File: Pinta/Dialogs/DocumentPropertiesDialog.cs (hunk 6 above)
Comment (horkana) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: The Error dialog class uses Stetic which compounds the feeling that I should have avoided trying to hand write GUI code in the first place.

Comment (cameronwhite) [inline Pinta/Dialogs/DocumentPropertiesDialog.cs]: [It doesn't anymore](https://github.com/PintaProject/Pinta/blob/master/Pinta/Dialogs/ErrorDialog.cs#L79), but you probably haven't pulled in those changes yet.

**Comment by aggsol** [comment]
Is this merged anytime?

**Comment by cameronwhite** [comment]
I've merged the .ora bug fixes in dcb5c7af and 632c5660
